<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Spatial Biology Workshop - Downstream Analyses</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href=".././src/icon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src=".././src/icon.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Spatial Biology Workshop</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../pages/index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../pages/setup.html">
 <span class="menu-text">1. Setup</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../pages/spatial-data.html">
 <span class="menu-text">2. Load Spatial Data</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../pages/quality-control.html">
 <span class="menu-text">3. Quality Control</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../pages/downstream.html" aria-current="page">
 <span class="menu-text">4. Spatial Analyses</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/paganilab"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/LabPagani"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#objectives" id="toc-objectives" class="nav-link active" data-scroll-target="#objectives">Objectives ðŸŽ¯</a></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#spatially-variable-genes-with-giotto" id="toc-spatially-variable-genes-with-giotto" class="nav-link" data-scroll-target="#spatially-variable-genes-with-giotto">Spatially-variable genes with <code>Giotto</code></a></li>
  <li><a href="#check-out-spatial-modules-of-expression" id="toc-check-out-spatial-modules-of-expression" class="nav-link" data-scroll-target="#check-out-spatial-modules-of-expression">Check out spatial modules of expression</a></li>
  <li><a href="#spatial-domain-identification-with-spatially-informed-clustering" id="toc-spatial-domain-identification-with-spatially-informed-clustering" class="nav-link" data-scroll-target="#spatial-domain-identification-with-spatially-informed-clustering">Spatial domain identification with spatially-informed clustering</a></li>
  <li><a href="#ligand-receptor-activity-inference-with-cellchat" id="toc-ligand-receptor-activity-inference-with-cellchat" class="nav-link" data-scroll-target="#ligand-receptor-activity-inference-with-cellchat">Ligand-receptor activity inference with <code>CellChat</code></a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Downstream Analyses</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="objectives" class="level2">
<h2 class="anchored" data-anchor-id="objectives">Objectives ðŸŽ¯</h2>
<ul>
<li><em>Identification of spatially-coordinated genes</em></li>
<li><em>Identification of spatial domains</em></li>
<li><em>Ligand-receptor inference analysis with <code>CellChat</code></em></li>
</ul>
</section>
<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p>Now you will start to explore ways to use <strong>spatial information</strong> to gather conclusions from our sample. You might be thinking of questions like:</p>
<ul>
<li><em>Are there genes that display a specific pattern of expression?</em></li>
<li><em>Are my clusters more or less spatially organized? Can we include spatial information while clustering spots?</em></li>
<li><em>What about cell-cell interactions across spots? Are there ways to check them out in spatial data?</em></li>
</ul>
<p>Youâ€™ll see in this part that most of these questions are readily answered after a few steps given our properly built dataset!</p>
</section>
<section id="spatially-variable-genes-with-giotto" class="level2">
<h2 class="anchored" data-anchor-id="spatially-variable-genes-with-giotto">Spatially-variable genes with <code>Giotto</code></h2>
<p>As we have seen during the introduction, geographical data analysis approaches have been around for decades in the fields of economics and ecology. Ever since spatial profiling technologies were developed, researchers have re-purposed these mathematical approaches to study biological phenomena! One of the first applications initially developed for <em>smFISH</em> data was to check whether genes/probes displayed a more or less <strong>spatially organized pattern</strong> in a statistically relevant manner. The <a href="https://drieslab.github.io/Giotto_website/"><code>Giotto</code> package</a> takes advantage of these methods to detect <em>spatially-variable genes</em> in ST data.</p>
<p>Before that tough, we need a few setup steps. We first convert our <code>SeuratObject</code> into one which is compatible with the functionalities of <code>Giotto</code>. This is performed with a convenience function as shown below:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>gobj <span class="ot">&lt;-</span> <span class="fu">seuratToGiottoV5</span>(sp, <span class="at">spatial_assay =</span> <span class="st">'Visium10X'</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Need to re-scale the data after transformation into giotto object</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>gobj <span class="ot">&lt;-</span> <span class="fu">normalizeGiotto</span>(gobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>ðŸ’¡ Just write <code>gobj</code> in your console to inspect the <code>Giotto</code> object, everything we had before should be there!</p>
</blockquote>
<p>Then, we need to represent connections between spots in a way that the computer can understand, this is performed via <strong>networks</strong>. Think of them as ways to store connection between cells. Use the code below to create a <a href="https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm"><em>k</em>-Nearest Neighbor</a> graph where each spot will be connected to its <em>k</em> sorrounding neighbors.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create spatial network</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>gobj <span class="ot">&lt;-</span> <span class="fu">createSpatialNetwork</span>(<span class="at">gobject =</span> gobj,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">method =</span> <span class="st">"kNN"</span>, </span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">k =</span> <span class="dv">6</span>,</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">maximum_distance_knn =</span> <span class="dv">300</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">name =</span> <span class="st">"spatial_network"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are many ways to create spatial networks and the <em>k</em>-NN algorithm is only one of them, others include <em>radius-based approaches</em> and the <a href="https://en.wikipedia.org/wiki/Delaunay_triangulation">Delaunay triangulation</a>. Once we have created the network, we inspect it trying to see whether connections between spots make sense biologically! (i.e.&nbsp;we are not connecting far away spots unlikely to interact with wach other)</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot spatial network</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">spatPlot2D</span>(<span class="at">gobject =</span> gobj,  </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">show_network=</span> <span class="cn">TRUE</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">network_color =</span> <span class="st">"blue"</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">point_size =</span> <span class="dv">1</span>, </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">spatial_network_name =</span> <span class="st">"spatial_network"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can now go ahead and detect spatially-variable genes in the data. <code>Giotto</code> implements this functionality with a function called <code>binSpect</code>, which is based on gene expression binarization and <code>k-means</code> clustering. The method tries to identify genes which are either expressed or not in nearby cells connected in the network.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Spatially variable genes detection</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>ranktest <span class="ot">&lt;-</span> <span class="fu">binSpect</span>(gobj, </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bin_method =</span> <span class="st">"rank"</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">calc_hub =</span> <span class="cn">TRUE</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">subset_feats =</span> <span class="fu">VariableFeatures</span>(sp),</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">hub_min_int =</span> <span class="dv">5</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>                     <span class="at">spatial_network_name =</span> <span class="st">"spatial_network"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Q1.</strong> Inspect the resulting object, what are the top 3 most <em>spatially variable</em> genes?</p>
<p><strong>Q2.</strong> Plot them on the sample like shown below (<em>hint</em>: use the <code>spatFeatPlot2D</code> function)</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="check-out-spatial-modules-of-expression" class="level2">
<h2 class="anchored" data-anchor-id="check-out-spatial-modules-of-expression">Check out spatial modules of expression</h2>
<p>Another functionality of <code>Giotto</code> entails the discovery of <em>modules</em> of genes displaying coordinated and coherent expression patterns, these can be thought to encode a spatially coordinated biological function. As this is a computationally-intensive process, we restrict the search of modules to include the top 250 spatially-variable genes like shown below:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster the top 500 spatially variable genes into modules</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>ext_spatial_genes <span class="ot">&lt;-</span> ranktest[<span class="dv">1</span><span class="sc">:</span><span class="dv">250</span>,]<span class="sc">$</span>feats</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With the function below, we are computing which genes are more likely to compose modules by checking their cross-correlation profile.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute gene-gene spatial correlation</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>spat_cor_netw_DT <span class="ot">&lt;-</span> <span class="fu">detectSpatialCorFeats</span>(</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    gobj,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">method =</span> <span class="st">"network"</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">spatial_network_name =</span> <span class="st">"spatial_network"</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset_feats =</span> ext_spatial_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then cluster the features into 5 modules, this is an arbitrary choice and can be dependent on the nature of the analysis.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>spat_cor_netw_DT <span class="ot">&lt;-</span> <span class="fu">clusterSpatialCorFeats</span>(spat_cor_netw_DT, </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">name =</span> <span class="st">"spat_netw_clus"</span>, </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">k =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plot the clusters!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmSpatialCorFeats</span>(gobj,</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">spatCorObject =</span> spat_cor_netw_DT,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">use_clus_name =</span> <span class="st">"spat_netw_clus"</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">title =</span> <span class="cn">NULL</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can additionally summarize the activity of spatial modules into <strong>metascores</strong> at single spot level, this is equivalent to calculating a <em>signature score at the single-spot level</em> starting from th e lists of genes composing the modules. We expect to see a spatial patterning in the data. We can additionally plot a figure to check out how many genes there are per module and how much they are intercorrelated!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Rank clusters (also plots cluster size vs rank dotplot)</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>netw_ranks <span class="ot">&lt;-</span> <span class="fu">rankSpatialCorGroups</span>(</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  gobj,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">spatCorObject =</span> spat_cor_netw_DT, </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">use_clus_name =</span> <span class="st">"spat_netw_clus"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>cluster_genes_DT <span class="ot">&lt;-</span> <span class="fu">showSpatialCorFeats</span>(spat_cor_netw_DT, </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">use_clus_name =</span> <span class="st">"spat_netw_clus"</span>, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">show_top_feats =</span> <span class="dv">1</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Get gene-clusters associations</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>cluster_genes <span class="ot">&lt;-</span> cluster_genes_DT<span class="sc">$</span>clus </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cluster_genes) <span class="ot">&lt;-</span> cluster_genes_DT<span class="sc">$</span>feat_ID</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Create metagenes based on modules</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>gobj <span class="ot">&lt;-</span> <span class="fu">createMetafeats</span>(gobj, </span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>                                <span class="at">feat_clusters =</span> cluster_genes, </span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>                                <span class="at">name =</span> <span class="st">"cluster_metagene"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Plot metafeatures like shown below (<em>hint</em>: use the <code>spatCellPlot</code> function from the <code>Giotto</code> package but specify <code>netw_ranks$clusters</code> in the <code>cell_annotation_value</code> option)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also compute a set of the 30 top-expressed genes per module to then use to subset out sample and perform spatially-informed clustering!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get a list of the top 30 co-expressed genes per cluster - hide this and have them work it out</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>coexpr_dt <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">data.table</span>(</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">genes =</span> <span class="fu">names</span>(spat_cor_netw_DT<span class="sc">$</span>cor_clusters<span class="sc">$</span>spat_netw_clus),</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster =</span> spat_cor_netw_DT<span class="sc">$</span>cor_clusters<span class="sc">$</span>spat_netw_clus)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">setorder</span>(coexpr_dt, cluster)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>top30_coexpr_dt <span class="ot">&lt;-</span> coexpr_dt[, <span class="fu">head</span>(.SD, <span class="dv">30</span>) , by <span class="ot">=</span> cluster]</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>spatial_genes <span class="ot">&lt;-</span> top30_coexpr_dt<span class="sc">$</span>genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Q3</strong> Which are the top 3 genes for each spatial co-expression cluster?</p>
</section>
<section id="spatial-domain-identification-with-spatially-informed-clustering" class="level2">
<h2 class="anchored" data-anchor-id="spatial-domain-identification-with-spatially-informed-clustering">Spatial domain identification with spatially-informed clustering</h2>
<p>Now that we have a set of genes that we know display spatial patterns of expression, we can use these to group together cells with clustering like we have previously done. In this case, the results that we obtain will intrinsically retain spatial information, meaning that we can identify <em>spatially-guided</em> results. We therefore start from the spatial genes to calculate again the principal components, umap, network and clustering!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform PCA on features from spatially-variable genes</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">RunPCA</span>(sp,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> spatial_genes,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction.name =</span> <span class="st">"custom_pca"</span>,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-compute UMAP embedding</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindNeighbors</span>(sp, <span class="at">reduction=</span><span class="st">'custom_pca'</span>, <span class="at">graph.name=</span><span class="st">'custom_NN'</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co">#sp &lt;- Seurat::RunUMAP(sp, graph='custom_NN', reduction.name='custom_UMAP')</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster spots based on PCA neighbors from spatially-variable genes</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">FindClusters</span>(sp, </span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>                            <span class="at">graph.name =</span> <span class="st">"custom_NN"</span>,</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>                            <span class="at">cluster.name =</span> <span class="st">"custom_clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 1339
Number of edges: 12730

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.7970
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
</div>
</div>
<p><strong>Q4.</strong> Display the resulting spatial domains you obtained (<em>hint</em>: try using the <code>SpatialDimPlot</code> function with the <code>custom_leiden</code> metadata field)</p>
<p><strong>Q5.</strong> Calculate the results of the previous clusters with the clustering obtained now, how much is each previous cluster contained in the new ones?</p>
<p><strong>Q6.</strong> Display cell type proportions across the identified domains using a heatmap (<em>hint</em>: see the <code>pheatmap</code> package)</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="ligand-receptor-activity-inference-with-cellchat" class="level2">
<h2 class="anchored" data-anchor-id="ligand-receptor-activity-inference-with-cellchat">Ligand-receptor activity inference with <code>CellChat</code></h2>
<p>Ever since the widespread generation of scRNA-seq data, computational methods to infer cell-cell interaction events (CCI) have been developed based on co-expression of ligand- and receptor-encoding genes across clusters. Spatial techonologies deepen this analysis by giving us the chance to weight the inferred interaction by their actual spatial proximity, limiting the chance of identifying false positive results.</p>
<p>The method that we will exploit in this part of the workshop is packaged into the <a href="https://github.com/jinworks/CellChat"><code>CellChat</code> framework</a>.</p>
<p>We first take the original <code>SeuratObject</code> and create a <code>CellChat</code> object with its own class. We first isolate raw data and import scale factors since <code>CellChat</code> takes into account spatial information, check out the steps below:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get normalized data matrix</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>data.input <span class="ot">=</span> Seurat<span class="sc">::</span><span class="fu">GetAssayData</span>(sp, <span class="at">slot =</span> <span class="st">"data"</span>, <span class="at">assay =</span> <span class="st">"Visium10X"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Source scale factors for um to px conversion</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Of note, the 'spot_diameter_fullres' factor is different from the `spot` in Seurat object and thus users still need to get the value from the original json file. </span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>scalefactors <span class="ot">=</span> jsonlite<span class="sc">::</span><span class="fu">fromJSON</span>(<span class="at">txt =</span> <span class="fu">file.path</span>(<span class="st">"./data/spatial"</span>, <span class="st">'scalefactors_json.json'</span>))</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># manually create a dataframe consisting of the cell labels, make sure to avoid 0 in the labels, so we create another column</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">orig.labels =</span> <span class="fu">Idents</span>(sp), <span class="at">samples =</span> <span class="st">"sample1"</span>, <span class="at">row.names =</span> <span class="fu">names</span>(<span class="fu">Idents</span>(sp))) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">labels =</span> <span class="fu">paste0</span>(<span class="st">'cluster-'</span>,orig.labels))</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>samples <span class="ot">&lt;-</span> <span class="fu">factor</span>(meta<span class="sc">$</span>samples)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(meta<span class="sc">$</span>labels) <span class="co"># check the cell labels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "cluster-3" "cluster-1" "cluster-4" "cluster-5" "cluster-8" "cluster-0"
[7] "cluster-6" "cluster-7" "cluster-2"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>spatial.locs <span class="ot">=</span> <span class="fu">GetTissueCoordinates</span>(sp, <span class="at">scale =</span> <span class="cn">NULL</span>, <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"imagerow"</span>, <span class="st">"imagecol"</span>))[,<span class="fu">c</span>(<span class="st">'x'</span>,<span class="st">'y'</span>)] <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>spot.size <span class="ot">=</span> <span class="dv">65</span> <span class="co"># the theoretical spot size (um) in 10X Visium</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>conversion.factor <span class="ot">=</span> spot.size<span class="sc">/</span>scalefactors<span class="sc">$</span>spot_diameter_fullres</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>spatial.factors <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">ratio =</span> conversion.factor, <span class="at">tol =</span> spot.size<span class="sc">/</span><span class="dv">2</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>d.spatial <span class="ot">&lt;-</span> <span class="fu">computeCellDistance</span>(<span class="at">coordinates =</span> spatial.locs, <span class="at">ratio =</span> spatial.factors<span class="sc">$</span>ratio, <span class="at">tol =</span> spatial.factors<span class="sc">$</span>tol)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(d.spatial[d.spatial<span class="sc">!=</span><span class="dv">0</span>]) <span class="co"># this value should approximately equal 100um for 10X Visium data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 99.57156</code></pre>
</div>
</div>
<p>Now create the <code>CellChat</code> object and store it in the <code>cellchat</code> variable:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create CellChat object</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">createCellChat</span>(<span class="at">object =</span> data.input, <span class="at">meta =</span> meta, <span class="at">group.by =</span> <span class="st">"labels"</span>,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">datatype =</span> <span class="st">"spatial"</span>, <span class="at">coordinates =</span> spatial.locs, <span class="at">spatial.factors =</span> spatial.factors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Create a CellChat object from a data matrix"
Create a CellChat object from spatial transcriptomics data... 
Set cell identities for the new CellChat object 
The cell groups used for CellChat analysis are  cluster-0, cluster-1, cluster-2, cluster-3, cluster-4, cluster-5, cluster-6, cluster-7, cluster-8 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's see it</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>cellchat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class CellChat created from a single dataset 
 33538 genes.
 1339 cells. 
CellChat analysis of spatial data! The input spatial locations are 
                   x_cent y_cent
AAACAAGTATCTCCCA-1   6982   7857
AAACCGGGTAGGTACC-1   6061   3067
AAACCGTTCGTCCAGG-1   7192   3969
AAACCTCATGAAGTTG-1   5496   2486
AAACGAGACGGTTGAT-1   5286   6373
AAACTGCTGGCTCCAA-1   6410   5592</code></pre>
</div>
</div>
<p>Build a ligand-receptor database:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>CellChatDB <span class="ot">&lt;-</span> CellChatDB.human <span class="co"># use CellChatDB.human since running on human data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Take a look at it:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">glimpse</span>(CellChatDB<span class="sc">$</span>interaction)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 3,233
Columns: 28
$ interaction_name         &lt;chr&gt; "TGFB1_TGFBR1_TGFBR2", "TGFB2_TGFBR1_TGFBR2",â€¦
$ pathway_name             &lt;chr&gt; "TGFb", "TGFb", "TGFb", "TGFb", "TGFb", "TGFbâ€¦
$ ligand                   &lt;chr&gt; "TGFB1", "TGFB2", "TGFB3", "TGFB1", "TGFB1", â€¦
$ receptor                 &lt;chr&gt; "TGFbR1_R2", "TGFbR1_R2", "TGFbR1_R2", "ACVR1â€¦
$ agonist                  &lt;chr&gt; "TGFb agonist", "TGFb agonist", "TGFb agonistâ€¦
$ antagonist               &lt;chr&gt; "TGFb antagonist", "TGFb antagonist", "TGFb aâ€¦
$ co_A_receptor            &lt;chr&gt; "", "", "", "", "", "", "", "", "", "", "", "â€¦
$ co_I_receptor            &lt;chr&gt; "TGFb inhibition receptor", "TGFb inhibition â€¦
$ annotation               &lt;chr&gt; "Secreted Signaling", "Secreted Signaling", "â€¦
$ interaction_name_2       &lt;chr&gt; "TGFB1 - (TGFBR1+TGFBR2)", "TGFB2 - (TGFBR1+Tâ€¦
$ evidence                 &lt;chr&gt; "KEGG: hsa04350", "KEGG: hsa04350", "KEGG: hsâ€¦
$ is_neurotransmitter      &lt;chr&gt; "FALSE", "FALSE", "FALSE", "FALSE", "FALSE", â€¦
$ ligand.symbol            &lt;chr&gt; "TGFB1", "TGFB2", "TGFB3", "TGFB1", "TGFB1", â€¦
$ ligand.family            &lt;chr&gt; "TGF-beta", "TGF-beta", "TGF-beta", "TGF-betaâ€¦
$ ligand.location          &lt;chr&gt; "Extracellular matrix, Secreted, Extracellulaâ€¦
$ ligand.keyword           &lt;chr&gt; "Disease variant, Signal, Reference proteome,â€¦
$ ligand.secreted_type     &lt;chr&gt; "growth factor", "growth factor", "cytokine;gâ€¦
$ ligand.transmembrane     &lt;chr&gt; "FALSE", "FALSE", "TRUE", "FALSE", "FALSE", "â€¦
$ receptor.symbol          &lt;chr&gt; "TGFBR2, TGFBR1", "TGFBR2, TGFBR1", "TGFBR2, â€¦
$ receptor.family          &lt;chr&gt; "Protein kinase superfamily, TKL Ser/Thr protâ€¦
$ receptor.location        &lt;chr&gt; "Cell membrane, Secreted, Membrane raft, Cellâ€¦
$ receptor.keyword         &lt;chr&gt; "Membrane, Secreted, Disulfide bond, Kinase, â€¦
$ receptor.surfaceome_main &lt;chr&gt; "Receptors", "Receptors", "Receptors", "Recepâ€¦
$ receptor.surfaceome_sub  &lt;chr&gt; "Act.TGFB;Kinase", "Act.TGFB;Kinase", "Act.TGâ€¦
$ receptor.adhesome        &lt;chr&gt; "", "", "", "", "", "", "", "", "", "", "", "â€¦
$ receptor.secreted_type   &lt;chr&gt; "", "", "", "", "", "", "", "", "", "", "", "â€¦
$ receptor.transmembrane   &lt;chr&gt; "TRUE", "TRUE", "TRUE", "TRUE", "TRUE", "TRUEâ€¦
$ version                  &lt;chr&gt; "CellChatDB v1", "CellChatDB v1", "CellChatDBâ€¦</code></pre>
</div>
</div>
<ul>
<li>How many ligand-receptor pairs in the database refer to secretory signalling pathways? (<em>hint</em>: use the <code>showDatabaseCategory</code> function from <code>CellChat</code>)</li>
</ul>
<p>We can now move into the actual <code>CellChat</code> main workflow steps:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter database of interactions to keep Secreted Signaling only</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>CellChatDB.use <span class="ot">&lt;-</span> <span class="fu">subsetDB</span>(CellChatDB, <span class="at">search =</span> <span class="st">"Secreted Signaling"</span>, <span class="at">key =</span> <span class="st">"annotation"</span>) </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Insert the filtered database into the cellchat object</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>cellchat<span class="sc">@</span>DB <span class="ot">&lt;-</span> CellChatDB.use</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co"># This step is necessary even if using the whole database, let's keep the highly variable genes in the data for speed</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">subsetData</span>(cellchat, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(sp))</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="co"># With these functions it starts computing interaction scores in each spot</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">identifyOverExpressedGenes</span>(cellchat)</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">identifyOverExpressedInteractions</span>(cellchat, <span class="at">variable.both =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The number of highly variable ligand-receptor pairs used for signaling inference is 46 </code></pre>
</div>
</div>
<p>Now we can compute interaction probabilities across clusters based on spatial proximity:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute communication probabilities across the originally identified cluster</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">computeCommunProb</span>(cellchat, <span class="at">type =</span> <span class="st">"truncatedMean"</span>, <span class="at">trim =</span> <span class="fl">0.1</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">distance.use =</span> <span class="cn">TRUE</span>, <span class="at">interaction.range =</span> <span class="dv">250</span>, <span class="at">scale.distance =</span> <span class="fl">0.01</span>,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">contact.dependent =</span> <span class="cn">TRUE</span>, <span class="at">contact.range =</span> <span class="dv">100</span>, <span class="at">nboot=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>truncatedMean is used for calculating the average gene expression per cell group. 
[1] "&gt;&gt;&gt; Run CellChat on spatial transcriptomics data using distances as constraints of the computed communication probability &lt;&lt;&lt; [2025-09-09 02:35:57.822061]"
Molecules of the input L-R pairs are diffusible. Run CellChat in a diffusion manner based on the `interaction.range`.
[1] "&gt;&gt;&gt; CellChat inference is done. Parameter values are stored in `object@options$parameter` &lt;&lt;&lt; [2025-09-09 02:36:47.420939]"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregate results</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">aggregateNet</span>(cellchat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Visualize the number of interactions detected and their strenght using <em>chord diagrams</em> like shown below (<em>hint</em>: use the <code>netVisual_circle</code> function)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">xpd=</span><span class="cn">TRUE</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of interactions</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_circle</span>(cellchat<span class="sc">@</span>net<span class="sc">$</span>count, <span class="at">vertex.weight =</span> <span class="fu">rowSums</span>(cellchat<span class="sc">@</span>net<span class="sc">$</span>count), <span class="at">weight.scale =</span> T, <span class="at">label.edge=</span> F, <span class="at">title.name =</span> <span class="st">"Number of interactions"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Interaction strength</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_circle</span>(cellchat<span class="sc">@</span>net<span class="sc">$</span>weight, <span class="at">vertex.weight =</span> <span class="fu">rowSums</span>(cellchat<span class="sc">@</span>net<span class="sc">$</span>weight), <span class="at">weight.scale =</span> T, <span class="at">label.edge=</span> F, <span class="at">title.name =</span> <span class="st">"Interaction weights/strength"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Visualize pairwise interactions using a heatmap (<em>hint</em>: use the <code>netVisual_heatmap</code> function)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_heatmap</span>(cellchat, <span class="at">measure =</span> <span class="st">"count"</span>, <span class="at">color.heatmap =</span> <span class="st">"Blues"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>Q7.</strong> Which are the clusters with the highest level of interactions? (<em>hint</em>: check out the <code>cellchatnet$count</code> object)</p>
<p><strong>Q8.</strong> Which are the top 3 interactions across <code>cluster-3</code> (Tumor) and <code>cluster-2</code> (Stroma/Fibroblasts)? (<em>hint</em>: use the <code>subsetCommunication</code> function from <code>CellChat</code>)</p>
<p>Letâ€™s compute the overall activity of CCI at the <strong>pathway level</strong>, in this case the goal is to summarize the interactions that we have retrieved to ones that pertain to specific intracellular pathways in order to get a sense of what is going on in communicating cells!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>cellchat <span class="ot">&lt;-</span> <span class="fu">computeCommunProbPathway</span>(cellchat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Letâ€™s now see the activity of a specific pathway always using a <em>chord diagram</em>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>pathways.show <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"FGF"</span>) </span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Circle plot</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">xpd =</span> <span class="cn">TRUE</span>) </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="fu">netVisual_aggregate</span>(cellchat, <span class="at">signaling =</span> pathways.show, <span class="at">layout =</span> <span class="st">"circle"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>Q9.</strong> Plot this in space as well like shown below (<em>hint</em>: use the <code>netVisual_aggregate</code> function)</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can additionally plot a â€œbinarizedâ€ sender/receiver version of the plot above for a specific pathway:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take an input of a ligand-receptor pair and show expression in binary</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">spatialFeaturePlot</span>(cellchat, <span class="at">pairLR.use =</span> <span class="st">"FGF18_FGFR4"</span>, <span class="at">point.size =</span> <span class="dv">1</span>, <span class="at">do.binary =</span> <span class="cn">TRUE</span>, <span class="at">cutoff =</span> <span class="fl">0.05</span>, <span class="at">enriched.only =</span> F, <span class="at">color.heatmap =</span> <span class="st">"Reds"</span>, <span class="at">direction =</span> <span class="dv">1</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">spatialFeaturePlot</span>(cellchat, <span class="at">pairLR.use =</span> <span class="st">"CXCL9_ACKR1"</span>, <span class="at">point.size =</span> <span class="dv">1</span>, <span class="at">do.binary =</span> <span class="cn">TRUE</span>, <span class="at">cutoff =</span> <span class="fl">0.05</span>, <span class="at">enriched.only =</span> F, <span class="at">color.heatmap =</span> <span class="st">"Reds"</span>, <span class="at">direction =</span> <span class="dv">1</span>)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="downstream_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Congratulations! You made it all the way to the end of the workshop! ðŸ¥³</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
      <div class="nav-footer-center">Copyright 2025, Mattia Toninelli</div>
  </div>
</footer>



</body></html>