<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Spatial Biology Workshop - Quality Control</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href=".././src/icon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src=".././src/icon.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Spatial Biology Workshop</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../pages/index.html">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../pages/setup.html">
 <span class="menu-text">1. Setup</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../pages/spatial-data.html">
 <span class="menu-text">2. Load Spatial Data</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../pages/quality-control.html" aria-current="page">
 <span class="menu-text">3. Quality Control</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../pages/downstream.html">
 <span class="menu-text">4. Spatial Analyses</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/paganilab"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/LabPagani"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#objectives" id="toc-objectives" class="nav-link active" data-scroll-target="#objectives">Objectives üéØ</a></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a>
  <ul class="collapse">
  <li><a href="#why-filter" id="toc-why-filter" class="nav-link" data-scroll-target="#why-filter">Why filter?</a></li>
  <li><a href="#special-considerations-with-spatial-data" id="toc-special-considerations-with-spatial-data" class="nav-link" data-scroll-target="#special-considerations-with-spatial-data">Special considerations with spatial data</a></li>
  </ul></li>
  <li><a href="#qc-thresholds" id="toc-qc-thresholds" class="nav-link" data-scroll-target="#qc-thresholds">QC Thresholds</a></li>
  <li><a href="#filtering" id="toc-filtering" class="nav-link" data-scroll-target="#filtering">Filtering</a></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">Normalization</a></li>
  <li><a href="#selection-of-highly-variable-features" id="toc-selection-of-highly-variable-features" class="nav-link" data-scroll-target="#selection-of-highly-variable-features">Selection of Highly-variable features</a></li>
  <li><a href="#clustering-spatially-unaware" id="toc-clustering-spatially-unaware" class="nav-link" data-scroll-target="#clustering-spatially-unaware">Clustering (spatially-unaware)</a></li>
  <li><a href="#which-will-be-the-dimensionality-of-our-data" id="toc-which-will-be-the-dimensionality-of-our-data" class="nav-link" data-scroll-target="#which-will-be-the-dimensionality-of-our-data">Which will be the ‚Äúdimensionality‚Äù of our data?</a></li>
  <li><a href="#non-linear-representation-of-the-data-umapt-sne" id="toc-non-linear-representation-of-the-data-umapt-sne" class="nav-link" data-scroll-target="#non-linear-representation-of-the-data-umapt-sne">Non-linear representation of the data (UMAP/t-SNE)</a></li>
  <li><a href="#grouping-spots-together-with-clustering" id="toc-grouping-spots-together-with-clustering" class="nav-link" data-scroll-target="#grouping-spots-together-with-clustering">Grouping spots together with clustering</a></li>
  <li><a href="#find-differentially-expressed-marker-genes" id="toc-find-differentially-expressed-marker-genes" class="nav-link" data-scroll-target="#find-differentially-expressed-marker-genes">Find differentially expressed marker genes</a></li>
  <li><a href="#cell-type-enrichment-of-spots" id="toc-cell-type-enrichment-of-spots" class="nav-link" data-scroll-target="#cell-type-enrichment-of-spots">Cell type enrichment of spots</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Quality Control</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="objectives" class="level2">
<h2 class="anchored" data-anchor-id="objectives">Objectives üéØ</h2>
<ul>
<li><em>Data normalization and filtering</em></li>
<li><em>Normalize and explore the data with functions provided by the <code>Seurat</code> and <code>Giotto</code> packages</em></li>
<li><em>Visual exploration of QC metrics over the sample</em></li>
</ul>
</section>
<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p>As in all bioinformatics analyses, we need to make sure that the data we are analyzing is as sound as possible. Biological data is especially noisy and chaotic by its very nature, hence it is important to perform proper quality control (<strong>QC</strong>) steps and filtering.</p>
<section id="why-filter" class="level3">
<h3 class="anchored" data-anchor-id="why-filter">Why filter?</h3>
<p>Filtering can help remove low-quality cells/spots that would otherwise introduce technical noise. For example, the presence of spots with lower numbers of features detected than other good quality spots <strong>can badly influence the clustering</strong> into groupings driven by total expression detected (UMIs) rather than by actual transcriptional content. Spots can also be considered low quality if they show signs of being poorly representative of the biological activity being studied. Some examples are cells that are actively dying (high mitochondrial gene expression) or observations that are actually doublets (in the case of single cell RNA-seq or imaging-based technologies). Removing these sources of non-relevant variation represents a key step for downstream analyses, <strong>improving the ratio of biological signal relative to technical noise</strong>.</p>
</section>
<section id="special-considerations-with-spatial-data" class="level3">
<h3 class="anchored" data-anchor-id="special-considerations-with-spatial-data">Special considerations with spatial data</h3>
<p>while filtering is beneficial in all usual single-cell analyses, spatial data requires an additional level of consideration. Removing a poor quality cell from a spatial network corresponds to removing a data point from its native environment.</p>
<blockquote class="blockquote">
<p>üö® If done without proper caution, this could damage our understanding of spatial niche makeup and cell-cell crosstalk!</p>
</blockquote>
</section>
</section>
<section id="qc-thresholds" class="level2">
<h2 class="anchored" data-anchor-id="qc-thresholds">QC Thresholds</h2>
<p>Filtering can be performed in many ways and using combinations of covariates to make sure we are retaining cells/spots we think are interesting for the sake of the analysis. Take note that this is a very analysis-based step! In our case, we are going to filter spots based on the amount of genes they encode and their total UMI content.</p>
<blockquote class="blockquote">
<p>üí° To get a feeling for what could be sensible thresholds to use for filtering, we have to get an idea of the average UMIs and genes!</p>
</blockquote>
<p><strong>Q1.</strong> Go ahead and compute the average UMI value and gene values per spot, what are these values?</p>
</section>
<section id="filtering" class="level2">
<h2 class="anchored" data-anchor-id="filtering">Filtering</h2>
<p>Once we have our filtering values sorted out, we need to actually filter our <em>bad</em> spots. Firstly though, since we are dealing with spatial data, we want to check whether the filtering we are applying could end up tampering with the original spatial organization of the sample!</p>
<p><strong>Q2.</strong> Determine a quantile-based approach to filter spots based on the <code>nCount_Visium10X</code> and <code>nFeature_Visium10X</code> columns, keeping only spots above the 10% percentile for both measures (let‚Äôs call them <code>genesq</code> for the number of genes and <code>umisq</code> for the number of counts, we‚Äôll use these later!) - <em>hint</em>: use the <code>quantile</code> function.</p>
<p><strong>Q3.</strong> Compute an additional <code>meta.data</code> column named <code>to_keep</code> which contains values <code>Keep</code> or <code>Discard</code>. If a spot passes QC then gets assigned a <code>Keep</code> else it is flagged by a <code>Discard</code>. (<em>hint</em>: check out the <code>mutate</code> and <code>ifelse</code> function to create the column)</p>
<p><strong>Q4.</strong> Plot the new column over the sample and check where spots that would be discarded are located</p>
<p><strong>Q5.</strong> How many spots would be filtered out?</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Once we are done, we can actually filter the sample!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">subset</span>(sp, <span class="at">subset =</span> to_keep <span class="sc">==</span> <span class="st">'Keep'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>üí° Did the filtering actually returned the expected amount of spots?</p>
</blockquote>
</section>
<section id="normalization" class="level2">
<h2 class="anchored" data-anchor-id="normalization">Normalization</h2>
<p>After the removal of unwanted spots from the dataset, it‚Äôs time to <em>normalize</em> it. What this step tries to achieve is the removal of unwanted differences across spots given by their varying <em>library sizes</em>. More specifically, we use the <code>NormalizeData</code> function from <code>Seurat</code> that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default) and log-transforms the result. Normalized expression values are then stored in <code>sp[["RNA"]]$data</code>.</p>
<ul>
<li>Use the <code>NormalizeData</code> function from <code>Seurat</code> to normalize the expression levels in our sample</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(sp, <span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>, <span class="at">scale.factor =</span> <span class="dv">10000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="selection-of-highly-variable-features" class="level2">
<h2 class="anchored" data-anchor-id="selection-of-highly-variable-features">Selection of Highly-variable features</h2>
<p>Now that we have our filtered dataset, we are ready to start working on the spots at the transcriptional level. First, we want to subset our <code>counts</code> matrix to retain information coming from genes which show interesting patterns in the data, in other words we want to keep genes which <em>vary</em> across our spots, since those should be the interesting ones!</p>
<p>To do so, try to implement the following steps in your code:</p>
<p><strong>Q6.</strong> Use the <code>FindVariableFeatures</code> function from the <code>Seurat</code> package to detect highly-variable genes in the data and keep the top 4,000 genes using the <code>vst</code> method</p>
<p><strong>Q7.</strong> Check which are the top 10 most variable genes in the data (<em>hint</em>: try a combination of the <code>head</code> and <code>VariableFeatures</code> function)</p>
<p><strong>Q8.</strong> Plot the names of the top 50 highly-variable genes like shown below (<em>hint</em>: use the <code>VariableFeaturePlot</code> function)</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Top 10 most highly variable genes?</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>top10 <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>üí° Selecting HVGs is really only a trick to enhance the signal coming from the inner variability of the data, we are not discarding any information as all genes are available to us all the time anyways!</p>
</blockquote>
</section>
<section id="clustering-spatially-unaware" class="level2">
<h2 class="anchored" data-anchor-id="clustering-spatially-unaware">Clustering (spatially-unaware)</h2>
<p>Clustering allows us to group spots together based on <em>transcriptional similarity</em>! In order to achieve this, we have to first perform a set of steps beforehand to prepare the data for the <strong>clustering algorithm</strong> we will use.</p>
<p>First, we apply a linear transformation (‚Äòscaling‚Äô) that is a standard pre-processing step prior to dimensionality reduction techniques like PCA. The <code>ScaleData()</code> function:</p>
<ul>
<li>Shifts the expression of each gene, so that the mean expression across cells is 0</li>
<li>Scales the expression of each gene, so that the variance across cells is 1
<ul>
<li>This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate</li>
</ul></li>
<li>The results of this are stored in pbmc[[‚ÄúRNA‚Äù]]$scale.data</li>
<li>By default, only variable features are scaled.</li>
</ul>
<p>Follow along and perform these steps in your code:</p>
<p>Let‚Äôs <em>scale</em> the data</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(sp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then perform <strong>linear</strong> <em>dimensionality reduction</em> with PCA. In this case we want to isolate the axes of variation in the data, for each cell we will have new ‚Äúmetafeatures‚Äù in the form of Principal Components (PCs) which are linear combinations (think of it as a summary) of the contribution of the original genes of that cell!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform PCA on HVGs</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(sp, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> sp))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Q9.</strong> How many PCs are computed by default?</p>
<p><strong>Q10.</strong> Which are the top 10 genes contributing to the first 3 PCs (based on their loadings, <em>hint</em>: use the <code>VizDimLoadings</code> function)?</p>
<p>Great! We can now plot our spots in PCA space:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot PC1 vs PC2</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(sp, <span class="at">reduction =</span> <span class="st">"pca"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>Q11.</strong> Now plot the PC score of each spot directly on the tissue in space (<em>hint</em>: this is stored in the <code>sp@reductions$pca</code> slot), to see whether there is a correlation between principal component and tissue location (like shown below), what do you expect?</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>üí° If you are interested in exploring the relationship between cells, PCs and genes, try out the <code>DimHeatmap</code> function!</p>
</blockquote>
</section>
<section id="which-will-be-the-dimensionality-of-our-data" class="level2">
<h2 class="anchored" data-anchor-id="which-will-be-the-dimensionality-of-our-data">Which will be the ‚Äúdimensionality‚Äù of our data?</h2>
<p>In order to overcome the technical noise in the data and highlight sources of relevant biological variation, we evaluate the transcriptional similarity of cells based on their ‚Äúsummarized‚Äù gene expression, in other words based on their principal components! But how can we decide how many components we should retain (out of the default ones)? 10? 15? Maybe 20? ü§î</p>
<p>In order to choose we can generate what is known as an <strong>Elbow plot</strong>, a heuristic method by which we select the number of PCs based on the <em>percentage of variance</em> explained by each one. In other words we retain PCs which ‚Äúsummarize‚Äù better the variability within the data, discarding ones which instead relate more to subtle technical variation.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(sp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>Q12.</strong> Draw an elbow plot yourself (<em>hint</em>: use the convenient <code>ElbowPlot</code> function)</p>
<ul>
<li>How many PCs will you choose based on the plot? Would it make sense to pick 10?</li>
</ul>
</section>
<section id="non-linear-representation-of-the-data-umapt-sne" class="level2">
<h2 class="anchored" data-anchor-id="non-linear-representation-of-the-data-umapt-sne">Non-linear representation of the data (UMAP/t-SNE)</h2>
<p>Linear representation of the data are very useful to perform interpretable dimensionality reduction, nevertheless one might want to <strong>visually represent</strong> (and only that!) the data in a meaningful manner that preserves the transcriptional cell-cell similarity. Methods like <a href="https://umap-learn.readthedocs.io/en/latest/">UMAP</a> (Uniform Manifold Approximation and Projection) aim to do exactly this, they group together trancriptionally similar cells in a 2D plane, though sacrificing the representation of the global data structure.</p>
<p>The first thing that we want to do is select the amount of PCs that we want to use to compute the transcriptional neighbors of each spot (i.e.&nbsp;other spots which have a very similar PCs profile). You should have figured this out by looking at the elbow plot above!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>ndims <span class="ot">&lt;-</span> <span class="dv">10</span> <span class="co"># Put here your number of chosen PCs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let‚Äôs compute the neighbors and UMAP embedding!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(sp, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span>ndims)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(sp, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span>ndims)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Q13.</strong> Plot the UMAP displaying the QC information we computed before (<code>nFeature_RNA</code>, <code>nCount_RNA</code> and <code>transcriptome.complexity</code>) using the <code>FeaturePlot</code> function</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>What do you think of the results? How does it compare to your expectation also based on your previous experience with UMAP applied to single-cell data? Where you expecting more or less heterogeneity and diversity?</li>
</ul>
</section>
<section id="grouping-spots-together-with-clustering" class="level2">
<h2 class="anchored" data-anchor-id="grouping-spots-together-with-clustering">Grouping spots together with clustering</h2>
<p>Here we will try to group spots together based on their transcriptional similarity using specific <strong>clustering algorithms</strong>! This procedure enables us to see whether we can <em>unsupervisedly</em> dissect the transcriptional differences within the data. In order to do so, we <em>partition</em> the underlying graph connecting similar spots to one another (which we built with the <code>FindNeighbors</code> function) into islands of spots that we term <strong>clusters</strong>.</p>
<blockquote class="blockquote">
<p>üí° Note that we are still not considering spatial information for this analysis! We are purely evaluating transcriptional similarity!</p>
</blockquote>
<p>One of the main parameters to consider here is the clustering <code>resolution</code>, this determines the granularity of our final clustering. In other words, higher resolution often leads to smaller and finer clusters while lower resolution values return broader clusters. The definition of this parameter is dependent on a series of factors including the type of data, its source and biology (i.e.&nbsp;a tumor vs.&nbsp;a more homogeneous sample) and our knowledge of the sample we are analyzing. Usually clustering results are also interpretable based on heuristic metrics used to evaluate <strong>stability</strong>, the <a href="https://en.wikipedia.org/wiki/Silhouette_(clustering)"><em>silhouette score</em></a> being one of them.</p>
<p>Let‚Äôs now find clusters of spots in our data!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(sp, <span class="at">resolution =</span> <span class="fl">0.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 1339
Number of edges: 42820

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8047
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
</div>
</div>
<p><strong>Q14.</strong> How many clusters do you get?</p>
<p><strong>Q15.</strong> How many spots do we have per cluster on average?</p>
<p><strong>Q16.</strong> Represent cluster identities on a UMAP and on a spatial plot, does this match your expectation? Do you see clusters also grouping spatially?</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>These are my cluster identities for the top 5 cells in the data:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>AAACAAGTATCTCCCA-1 AAACCGGGTAGGTACC-1 AAACCGTTCGTCCAGG-1 AAACCTCATGAAGTTG-1 
                 5                  4                  2                  1 
AAACGAGACGGTTGAT-1 
                 0 
Levels: 0 1 2 3 4 5 6 7 8</code></pre>
</div>
</div>
</section>
<section id="find-differentially-expressed-marker-genes" class="level2">
<h2 class="anchored" data-anchor-id="find-differentially-expressed-marker-genes">Find differentially expressed marker genes</h2>
<p>Once we have identified a grouping for our spots, we can move on to check which are the transcriptional differences across these groups. The goal of this procedure is to start to <em>get a feeling for the biology</em> behind the clusters we have identified. We can use the <code>FindAllMarkers</code> function to compute all marker genes for all spot clusters against all other spots at once. What this function does under the hood is to perform a statistical test (several are implemented in <code>Seurat</code>) for each gene and return the corresponding <code>p_val</code> and <code>log2FC</code> which describe the signficance and magnitude of the gene expression change in the categories we selected (i.e.&nbsp;cluster 1 <em>vs</em> all others).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># With `only.pos` we tell the software to only report genes with positive change in each cluster compared to all others</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(sp, <span class="at">only.pos =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Q17.</strong> Check out the top 5 markers for each cluster of spots, is there any marker you recognize?</p>
<p><strong>Q18.</strong> Plot the expression of one of the main markers across clusters like shown below (<em>hint</em>: use the <code>VlnPlot</code> function)</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>Q19.</strong> Plot the expression of one of the main markers on a UMAP embedding like shown below</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>Q20.</strong> Visualize the expression of the top 5 marker genes per cluster in a heatmap (<em>hint</em>: use the <code>DoHeatmap</code> function)</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>ü§î Based on the very technical nature of the technology, would it make sense to expect clear-cut marker genes for each cluster specifically related to cell types?</p>
</blockquote>
</section>
<section id="cell-type-enrichment-of-spots" class="level2">
<h2 class="anchored" data-anchor-id="cell-type-enrichment-of-spots">Cell type enrichment of spots</h2>
<p>As previously discussed, Visium assays gene expression in single larger-than-cells spots, meaning that each observation we get will contain mixed transcriptomes coming from neighboring cells. This can complicate downstream analysis when trying to infer cell-cell relationships and phenotypes. For this reason, there exist approaches aimed at dissecting this spot heterogeneity by either <strong>deconvolving</strong> the transcriptional signal of spots into contributions of single cell types or by evaluating the <strong>enrichment</strong> of cell type marker genes in the transcriptional profile of each single spot. The former are based on ‚Äúpure‚Äù transcriptional profiles usually distilled from a matched sample profiled with scRNA-seq, where cell clusters can be obtained and signals averaged. The latter approach is less computationally expensive, does not require scRNA-seq, but is entirely knowledge-based and less precise.</p>
<p>For the sake of this workshop, we will perform the second approach in the form of the <a href="https://igordot.github.io/clustermole/"><code>clustermole</code></a> package. This package contains a collection of cell-type specific marker genes coming from various databases whose activity can be avaluated across spots. In the background, we are essentially running a <code>GSEA</code> analysis on each single spot (<code>ssGSEA</code>) across all the available marker genesets in the database.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First extract normalized counts from our seurat object</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(sp) <span class="ot">&lt;-</span> <span class="st">"Visium10X"</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(<span class="at">object =</span> sp, <span class="at">layer =</span> <span class="st">'data'</span>) <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Run `clustermole`</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>enr <span class="ot">&lt;-</span> <span class="fu">clustermole_enrichment</span>(counts, <span class="at">species =</span> <span class="st">"hs"</span>, <span class="at">method =</span> <span class="st">'ssgsea'</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter results, keep only representative celltypes of interest OR celltypes in the `organ` Ovary</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>ct_to_match <span class="ot">=</span> <span class="fu">c</span>(<span class="st">'Macrophage'</span>,<span class="st">'CD4+ T cell'</span>, <span class="st">'CD8+ T cell'</span>, <span class="st">'Mesenchymal cell'</span>, <span class="st">'Epithelial cell'</span>, <span class="st">'Endothelial cell'</span>, <span class="st">'Fibroblast'</span>)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>enr_wide <span class="ot">&lt;-</span> enr <span class="sc">%&gt;%</span> <span class="fu">filter</span>(species <span class="sc">==</span> <span class="st">'Human'</span>) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(organ <span class="sc">==</span> <span class="st">'Ovary'</span> <span class="sc">|</span> celltype <span class="sc">%in%</span> ct_to_match) <span class="sc">%&gt;%</span> tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(.,<span class="at">id_cols=</span><span class="fu">c</span>(<span class="st">"cluster"</span>), <span class="at">names_from =</span> <span class="st">'celltype'</span>, <span class="at">values_from =</span> <span class="st">'score'</span>, <span class="at">values_fn =</span> mean, <span class="at">values_fill =</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"cluster"</span>)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a new column in metadata with labels from the enrichment (each spot is labelled for its highest score)</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>enr_wide<span class="sc">$</span>celltype <span class="ot">&lt;-</span> <span class="fu">names</span>(enr_wide)[<span class="fu">max.col</span>(enr_wide)]</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Add them to metadata</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>sp<span class="sc">@</span>meta.data <span class="ot">&lt;-</span> sp<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">merge</span>(enr_wide, <span class="at">by.y=</span><span class="dv">0</span>, <span class="at">by.x=</span><span class="dv">0</span>, <span class="at">all.x=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">"Row.names"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let‚Äôs now plot spatially the scores for specific gene signatures of interest like shown below!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot in space</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(sp, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">'Epithelial cell (Ovarian Cancer)'</span>), <span class="at">pt.size=</span><span class="dv">3</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">SpatialDimPlot</span>(sp,<span class="at">group.by =</span> <span class="st">'celltype'</span>, <span class="at">pt.size=</span><span class="dv">3</span>) <span class="sc">+</span> <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">'Paired'</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Draw a barplot using <code>ggplot2</code> highlighting the proportions of celltypes predicted across clusters in the sample</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="quality-control_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Based on the results above, can we confidently highlight a ‚Äútumor‚Äù cluster and a ‚Äústromal‚Äù one?</li>
</ul>
<p>Now we can move to additional downstream analyses which will also take into account spatial coordinates of spots on the sample!</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
      <div class="nav-footer-center">Copyright 2025, Mattia Toninelli</div>
  </div>
</footer>



</body></html>